pkg/api/jobs.go

```diff
@@ -138,6 +145,152 @@ func PrintJobsReport(w http.ResponseWriter, req *http.Request, currReport, twoDa
 		limit(req))
 }
 
+// PrintDBJobsReport renders a filtered summary of matching jobs.
+func PrintDBJobsReport(w http.ResponseWriter, req *http.Request,
+	dbc *db.DB, release string) {
+
+	var filter *Filter
+
+	queryFilter := req.URL.Query().Get("filter")
+	if queryFilter != "" {
+		filter = &Filter{}
+		if err := json.Unmarshal([]byte(queryFilter), filter); err != nil {
+			RespondWithJSON(http.StatusBadRequest, w, map[string]interface{}{"code": http.StatusBadRequest, "message": "Could not marshal query:" + err.Error()})
+			return
+		}
+	}
+
+	period := req.URL.Query().Get("period")
+	jobsResult, err := BuildJobResults(dbc, period, release)
+	if err != nil {
+		RespondWithJSON(http.StatusInternalServerError, w, map[string]interface{}{"code": http.StatusInternalServerError, "message": "Error building job report:" + err.Error()})
+		return
+	}
+
+	RespondWithJSON(http.StatusOK, w, jobsResult.
+		sort(req).
+		limit(req))
+}
+
+func BuildJobResults(dbc *db.DB, period, release string) (jobsAPIResult, error) {
+	now := time.Now()
+
+	// TODO: use actual start/num days settings from CLI once we understand what should
+	// be happening here, previous seems to include current. See PrepareStandardTestReports.
+	// Note that the CLI/most of the code says "start date" for what is actually the
+	// end of the date range, and does a walk back.
+	currDays := 7
+	prevDays := 14
+	if period == periodTwoDay {
+		currDays = 2
+		prevDays = 9 // 7 + 2
+	}
+	startDate := time.Now().Add(time.Duration(-1*prevDays*24) * time.Hour)
+	boundaryDate := time.Now().Add(time.Duration(-1*currDays*24) * time.Hour)
+	klog.Infof("BuildJobResult from %s -> %s -> %s", startDate, boundaryDate, time.Now())
+
+	// Load all ProwJobs and create a map of db ID to ProwJob:
+	var allJobs []models.ProwJob
+	dbr := dbc.DB.Find(&allJobs).Where("release = ?", release)
+	if dbr.Error != nil {
+		klog.Error(dbr)
+		return []apitype.Job{}, dbr.Error
+	}
+	allJobsByID := mapProwJobsByID(allJobs)
+	klog.V(3).Infof("Found %d jobs", len(allJobs))
+
+	// Load all ProwJobRuns in our period timestamp range:
+	var jobRunsInPeriod []models.ProwJobRun
+	dbr = dbc.DB.
+		Joins("JOIN prow_jobs ON prow_jobs.id = prow_job_runs.prow_job_id").
+		Where("timestamp BETWEEN ? AND ?", startDate, now).
+		Where("prow_jobs.release = ?", release).Find(&jobRunsInPeriod)
+	if dbr.Error != nil {
+		klog.Error(dbr)
+		return []apitype.Job{}, dbr.Error
+	}
+	klog.V(3).Infof("Found %d job runs in period between %s and %s",
+		len(jobRunsInPeriod),
+		startDate.Format(time.RFC3339),
+		now.Format(time.RFC3339))
+
+	// Build a map of job ID to the API job result we'll return.
+	// Iterate all job results, incrementing counters on the job result we'll return.
+	apiJobResults := map[uint]*apitype.Job{}
+	for _, jobRun := range jobRunsInPeriod {
```

Comment:
What do you think about doing more of this in SQL? We'd have to loop over jobs, but that's much smaller than job runs (and it might be possible to do that in sql too)

```sql
WITH results AS (
SELECT
	coalesce(count(case when succeeded = true AND timestamp BETWEEN NOW() - INTERVAL '14 DAY' AND NOW() - INTERVAL '7 DAY' then 1 end), 0) as previous_passes,
	coalesce(count(case when succeeded = false AND timestamp BETWEEN NOW() - INTERVAL '14 DAY' AND NOW() - INTERVAL '7 DAY' then 1 end), 0) as previous_failures,
	coalesce(count(case when timestamp BETWEEN NOW() - INTERVAL '14 DAY' AND NOW() - INTERVAL '7 DAY' then 1 end), 0) as previous_total_runs,
	coalesce(count(case when infrastructure_failure = true AND timestamp BETWEEN NOW() - INTERVAL '14 DAY' AND NOW() - INTERVAL '7 DAY' then 1 end), 0) as previous_infrastructure_failures,
	
	coalesce(count(case when succeeded = true AND timestamp > NOW() - INTERVAL '7 DAY' then 1 end), 0) as current_passes,
	coalesce(count(case when succeeded = false AND timestamp > NOW() - INTERVAL '7 DAY' then 1 end), 0) as current_failures,	
	coalesce(count(case when timestamp > NOW() - INTERVAL '7 DAY' then 1 end), 0) as current_total_runs,
	coalesce(count(case when infrastructure_failure = true AND timestamp > NOW() - INTERVAL '7 DAY' then 1 end), 0) as current_infrastructure_failures
FROM
	prow_job_runs JOIN prow_jobs ON prow_jobs.id = prow_job_runs.prow_job_id WHERE prow_jobs.name = 'promote-release-openshift-machine-os-content-e2e-aws-4.10'
)
SELECT *,
	previous_passes * 100.0 / previous_total_runs AS previous_pass_percentage,
	previous_failures * 100.0 / previous_total_runs AS previous_failure_percentage,
	current_passes * 100.0 / current_total_runs AS current_pass_percentage,
	current_failures * 100.0 / current_total_runs AS current_failure_percentage
FROM results;
```

Results:

```
-[ RECORD 1 ]--------------------+-----------------------
previous_passes                  | 591
previous_failures                | 4
previous_total_runs              | 595
previous_infrastructure_failures | 1
current_passes                   | 67
current_failures                 | 0
current_total_runs               | 67
current_infrastructure_failures  | 0
previous_pass_percentage         | 99.3277310924369748
previous_failure_percentage      | 0.67226890756302521008
current_pass_percentage          | 100.0000000000000000
current_failure_percentage       | 0.00000000000000000000
```

Reply:
We've got this working for all jobs on slack, will implement with a followup.